### PR TITLE
bpo-27099: IDLE - Convert built-in extensions to regular features

### PR DESCRIPTION
**Pull request by wohlganger**
* Adds fg color, bg color, font, and underline options for parens highlighting

https://bugs.python.org/issue27099

### CODE REVIEW

**Comment by wohlganger** [comment]
Per Terry J. Reedy recommendation, only highlighting style is changeable.

File: Lib/idlelib/help.html
Comment (terryjreedy) [inline Lib/idlelib/help.html]: help.html is generated from Doc/library/idle.rst.  So replace this edit with an edit of the latter.  I believe it should be straightforward.

**Comment by terryjreedy** [comment]
Charles, I cannot add commits to this issue, and a few others. Please follow the instructions at https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/#enabling-repository-maintainer-permissions-on-existing-pull-requests to make sure "Allow edits from maintainers." is checked for this and other IDLE issues. Please let me know what you find. If this permission is already granted, then I have some other problem with github.

**Comment by wohlganger** [comment]
Checked "Allow edits". Thank you for pointing it out; I'll remember to check it in the future.

**Comment by terryjreedy** [comment]
It is on now.  'Allow edits' is supposed to be checked by default.  If that changed, let me know and I will pass the information on to whoever can change it back.

**Comment by terryjreedy** [comment]
Because of the updates, this merged cleanly, so I added the news file.

I will worry about the IdleConfTest failures after looking at changes to other files.  Appeveyor also had this.

FAIL: test_keybinding (idlelib.idle_test.test_configdialog.KeysPageTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\projects\cpython\lib\idlelib\idle_test\test_configdialog.py", line 838, in test_keybinding
    {'AutoExpand_cfgBindings': {'expand-word': '<Key-F11>'}})
AssertionError: {} != {'AutoExpand_cfgBindings': {'expand-word': '<Key-F11>'}}
- {}
+ {'AutoExpand_cfgBindings': {'expand-word': '<Key-F11>'}}

**Comment by wohlganger** [comment]
The configdialog test is failing because it's expecting AutoExpand to be an extension, which it is not. In order for the test to be meaningful, there needs to be a default extension. I'm going to make ZoomHeight that default extension; it seems to be the ideal candidate.

File: Lib/idlelib/configdialog.py
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: Remove this addition and leave single spaced. 
Apparently done.

File: Lib/idlelib/configdialog.py
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: Remove. -- done

File: Lib/idlelib/configdialog.py
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: Remove.  Please don't go around reformatting, especially is a way that counters existing consistent style. -- apparently done here.

File: Lib/idlelib/configdialog.py
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: Revert this deletion.  'Extensions' key should become 'General' after widgets are moved there.
Reversion done. Name change later.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
configdialog.py

**Comment by wohlganger** [comment]
please wait before checking - I haven't fixed the entry for autocomplete and paragraph yet.

**Comment by wohlganger** [comment]
Ready to be checked

**Comment by wohlganger** [comment]
looking at the four failing tests, it wouldn't be difficult to fix them so that they pass with this patch, but the easy solution assumes there will always only be one extension - ZoomHeight. Future proofing it would be better, but I think that's outside the scope of this patch. Do you want me to fix those tests?

**Review by terryjreedy — APPROVED** [review]

File: Lib/idlelib/autocomplete.py
Comment (terryjreedy) [inline Lib/idlelib/autocomplete.py:38]: Why did you move this?  Although another issue, really, I can see moving within __init__ to have changes take effect  sooner, why make it conditional?

File: Lib/idlelib/autocomplete.py
Comment (terryjreedy) [inline Lib/idlelib/autocomplete.py]: One-liner is not worth a separate function.

File: Lib/idlelib/autocomplete.py
Comment (terryjreedy) [inline Lib/idlelib/autocomplete.py]: """Complete attribute name or file name.

Either on demand or after a user-selected delay after a key character, pop up a list of candidates.
"""
Note no return after initial quote.

File: Lib/idlelib/autoexpand.py
Comment (terryjreedy) [inline Lib/idlelib/autoexpand.py:13]: Leave 'There is only one instance of AutoExpand'.
This docstring actually belongs with the class, and will have to be moved when files are consolidated -- in another issue.

File: Lib/idlelib/codecontext.py
Comment (terryjreedy) [inline Lib/idlelib/codecontext.py]: This and other edits is an attempt to change the behavior of the feature, as opposed to simply changing the internal implementation.  These behavior changes have not been discussed, and mostly should be one or more separate issues.  Some of theme may already be so.

File: Lib/idlelib/config-extensions.def
Comment (terryjreedy) [inline Lib/idlelib/config-extensions.def:37]: As I said and explained on the issue, I want to leave non-key options in config-extensions.  So the 4 non-trivial [Xtension] sections should stay.  For former extensions, I would like to drop the requirement for an enable option when the entry is read in the editor.  Then we can delete the enable items.  I believe that this will be back-compatible with existing custom enable settings, but must verify.    (Ditto for not requiring some binding section.)

As I said yesterday, I now think the [Xtension_cfgBindings] should go, to be replaced by new entries in keysets.  This will require some change in IdleConf to deal better with missing entries in custom keysets.   But this was already needed, and perhaps should be a separate preliminary patch.

I think putting non-configurable key bindings in configuration files was a bit of a mistake.  In any case, these binding should be hard coded in the appropriate place.  At minimum, 'x_keys = idleConf(...)' should be replaced by 'x_keys = (current-idleconf-return).

File: Lib/idlelib/config-extensions.def
Comment (terryjreedy) [inline Lib/idlelib/config-extensions.def:63]: Keep

File: Lib/idlelib/config-extensions.def
Comment (terryjreedy) [inline Lib/idlelib/config-extensions.def:69]: Keep

File: Lib/idlelib/config-extensions.def
Comment (terryjreedy) [inline Lib/idlelib/config-extensions.def:78]: Keep.

File: Lib/idlelib/config-highlight.def
Comment (terryjreedy) [inline Lib/idlelib/config-highlight.def]: Delete parenmatch additions.

File: Lib/idlelib/config-keys.def
Comment (terryjreedy) [inline Lib/idlelib/config-keys.def]: Only current user-configurable 'event = keylist' pairs from [Xtension-cfgBindings] sections should be added.  These are
```
force-open-completions= <Control-Key-space>
expand-word= <Alt-Key-slash>
force-open-calltip= <Control-Key-backslash>
format-paragraph= <Alt-Key-q>
flash-paren= <Control-Key-0>
run-module= <Key-F5>
check-module= <Alt-Key-x>
zoom-height= <Alt-Key-2>
```
The keys for autocomplete, try-open-completions, try-open-calltip, refresh-calltip, and paren-closed are not and should not be configurable.  Adding new shortcuts, such as for toggle-code-context, will be a separate issue.

File: Lib/idlelib/config-main.def
Comment (terryjreedy) [inline Lib/idlelib/config-main.def]: Delete additions

File: Lib/idlelib/config.py
Comment (terryjreedy) [inline Lib/idlelib/config.py]: Delete parenmatch.

File: Lib/idlelib/config.py
Comment (terryjreedy) [inline Lib/idlelib/config.py]: Reduce additions to current user-configurable bindings.

File: Lib/idlelib/configdialog.py
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: delete

File: Lib/idlelib/configdialog.py
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: Move to General tab, though possible after I condense current stuff.

File: Lib/idlelib/configdialog.py
Comment (terryjreedy) [inline Lib/idlelib/configdialog.py]: delete

File: Lib/idlelib/editor.py
Comment (terryjreedy) [inline Lib/idlelib/editor.py]: Why here instead with other class imports?  Is there a circular import problem?

File: Lib/idlelib/editor.py
Comment (terryjreedy) [inline Lib/idlelib/editor.py]: Can you explain comment?

File: Lib/idlelib/editor.py
Comment (terryjreedy) [inline Lib/idlelib/editor.py]: Spaces around '=', as per PEP8.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
The overall problem is that the patch does more than specified on the issue. See the separate comments.

**Comment by terryjreedy** [comment]
The tests must pass one way or another.  I have thought about a possible dummy extension entry and file that would properly illustrate the features, but leaving zoomheight as is for now seems ok.  It might even be left as an (optional) extension indefinitely.  We could add comments to make it serve better as didactic example.

**Comment by terryjreedy** [comment]
After re-reading bpo-20580 and reading the sed commands in https://hg.python.org/cpython/rev/13826ff147e4, leaving zoomheight permanently an extension is not ok.  For reasons explained in the issue, Ned would like to get rid of the sed edits.

File: Lib/idlelib/autocomplete.py
Comment (wohlganger) [reply Lib/idlelib/autocomplete.py]: Built-in options are effective when they are changed from the config menu. Extensions require restarting IDLE. In order for these to function like built-ins, their changed settings needed to be reloaded from within the editor. That is what the reset function is for. There is one in each of the old extensions that had options. It seems the activation got lost when I was resolving conflicts, but it looks like the place to call it is now activate_config_changes in the configdialog. Would you rather I put the functionality back in, or that some of the new options behave differently than the old ones?

File: Lib/idlelib/editor.py
Comment (wohlganger) [reply Lib/idlelib/editor.py]: Travis or AppVeyor would error out with them when they were with the other imports, so I was pretty sure it was a circular import.

File: Lib/idlelib/editor.py
Comment (wohlganger) [reply Lib/idlelib/editor.py]: Code context and some of the other extensions depend on frame.text being packed, which occurs in color_config, called from ResetColorizer. If done before then, idle errors out.

File: Lib/idlelib/editor.py
Comment (terryjreedy) [reply Lib/idlelib/editor.py]: I am sure that some of the imports, such as the one for paragraph, could be moved to class level.  They should be tested one by one, and a comment added for those that must stay.   I will eventually do this is you don't.  I hope to eventually reduce the loopy dependencies, so I want to know which fail otherwise.  I also want to make sure the errors are from circular imports and not something else that should perhaps be fixed.

File: Lib/idlelib/autocomplete.py
Comment (terryjreedy) [reply Lib/idlelib/autocomplete.py]: I opened bpo-22707 two years ago for this problem, and intended that we work on it after this one. I just posted my analysis and made you nosy.  For reasons explained there, I want to leave the class attributes as class attributes.  Whether the idleConf wrapper are added here or in 22707, they should be called 'reload'; 'reset' implies changing back to initial values.  They should be classmethods called after the class statement.  This will preserve the current behavior, which I will manually test.

 The line length for ReformatParagraph should be a class attribute, and the test adjusted.  CodeContext().visible should be left alone until fixed or deleted.

I want to deal with changing code to update immediately in 22707, where the new code will be tested.  I will add more on that issue.

File: Lib/idlelib/editor.py
Comment (terryjreedy) [reply Lib/idlelib/editor.py]: To do import just once, on first instance: at class level, add for instance 'AutoComplete = None' and in __init__: 'if AutoComplete == None: self.do_imports()' where do_imports does delayed imports and sets class attributes.

**Comment by wohlganger** [comment]
It's getting unclear where you want the options. If they are left in config-extensions.def, they are automatically populated in the extensions tab. If there's the ability to change them from the General tab, users would expect them to be visible in .ilderc\config-main.cfg . Should I move them to config-main.def, or is there something I'm missing? Additionally - if we need the tests to pass AND we're moving zoomheight out of extensions, which causes the tests to fail - what is the plan there?

**Comment by terryjreedy** [comment]
I have consistently said that I want the widgets on the general tab.   I also want to leave then in config-extensions.def for back-compatibility. They will not appear on the extension tab. With 'enable' removed and with no binding, the truncated entries should not be recognized as legal extension entries.  We just need to make sure that they are ignored and no error or warning is emitted.  Sorry if I did not make this clear enough before.

Something similar is needed for the editor extension loading.  If nothing else, we can add an explicit exclusion list, but I don't expect to need that.  If you don't beat me to it, I will write both changes.

Users do not need to ever look at .idlerc/config-main.cfg.  It should only be edited with the dialog. We will add comments in the idlelib .def versions explaining the situation.

I plan on adding an annotated, functional, dummy extension, disabled by default.  It will be both an example of how to write an extension and a test subject.

**Comment by wohlganger** [comment]
I think I've gotten this patch how you want it. Since there's no dummy extension for testing, it still throws loads of errors in travis and appveyor.

**Comment by terryjreedy** [comment]
Good.  I will try to re-review in an hour or two.  I will add an extension and see what errors remain.

File: Lib/idlelib/autocomplete.py
Comment (terryjreedy) [inline Lib/idlelib/autocomplete.py]: Removing '=None' and the if statement caused subprocess startup to fail.
Fixed.

File: Lib/idlelib/autocomplete.py
Comment (terryjreedy) [inline Lib/idlelib/autocomplete.py]: I want these to be class methods called after the class is created.  I will fix for all 4 files. I also changed back to original args -- 'extensions', 'AutoComplete', 'popupwait'.

File: Lib/idlelib/config.py
Comment (terryjreedy) [inline Lib/idlelib/config.py]: Right code, wrong place.  For me, IDLE startup failed because I have customized AutoComplete.popupwait and therefore have an AutoComplete entry in config-extensions.cfg, which resulted in AutoComplete being added back to extns by the next 4 lines.  Problem fixed by moving exclusions after extns building is complete.

**Review by terryjreedy — COMMENTED** [review]
I could not work on this last night, but I am now.  Please to not push any more commits.  I had to make a couple of changes to get IDLE to run.  I will make other changes too.  When I get tests passing, I will push everything.

**Comment by terryjreedy** [comment]
I am not done: there are 2 errors left in text_config, + I want to reshuffle widgets on General tab.
+ more gracefully handling 'missing' entries in custom keysets.

**Review by terryjreedy — APPROVED** [review]

**Comment by terryjreedy** [comment]
There were more that two errors.  Anyway, I think this is complete enough.  There is a unittest for each new widget, and all tests pass for me.  I though the context highlight had moved to the highlights tabs, but if so, that disappeared.  I am willing to leave that to separate code context patch.  A method of dealing with the missing highlight in the user highlight file will be needed.  Ditto for dealing with missing keys in user key file.  

I still need to do manual lives test all the transported features.

**Comment by wohlganger** [comment]
I had moved codecontext and parenmatch options into the highlight tab and integrated them with the colorizer system, but you said you wanted them in general, so I moved them back. I think it'll make more sense to move them in a later patch than to make this any bigger than it needs to be, especially since it's passing tests now.

File: Lib/idlelib/config-keys.def
Comment (terryjreedy) [inline Lib/idlelib/config-keys.def:63]: On the Format menu, format-paragraph does not get its accelerator (hot-key).  In addition, `<<format-paragraph>>` does not appear in the ConfigDialog  keys list.  (The menu entry does work.)  The other 7 additions have their accelerators on the menu and event-seq pairs on the keys tab.

The reason was the absence of `<<format-paragraph>>` is from the core key list when I start with a custom key set.

File: Lib/idlelib/config.py
Comment (terryjreedy) [inline Lib/idlelib/config.py]: The additions, starting with `<<try-open-calltip>>`, include 7 of 8 user-configurable events, missing `<<format-paragraph>>`.   This prevented the needed substitution when using a custom keyset.  Fixed.

The additions also include 3 of 6 fixed events, missing <<autocomplete>>, <<refresh-call-tips>>, and <<toggle-code-context>>.  The purpose, I finally understand, was to bring about event_add.  But adding fixed events to the list of configurable pairs results in them appearing on the configure Keys page.  I will add explicit statements elsewhere.

File: Lib/idlelib/config-extensions.def
Comment (terryjreedy) [inline Lib/idlelib/config-extensions.def:41]: The '_bindings' entries had two effects:
1. event_add a virtual event associated with the event sequences.
2. bind the virtual event to an event handler (named after the event).
This new line in EditorWindow did the second:
`        text.bind("<<autocomplete>>", autocomplete.autocomplete_event)`
The lack of a replacement for 1. disabled the use of tab to open and tab tab to close a completion box.  Adding
`            self.text.event_add('<<autocomplete>>', '<Key-Tab>')`
in AutoComplete.__init__ fixed the problem.  I am thinking that both parts should be in the same file.  Either the event_add should be in EditorWindow or the binding should be in AutoComplete.  I am leaning toward the latter, but will decide tomorrow.

I decided to put all event_adds in EditorWindow for now.  I believe they only need to be called once in the process, not once for every window, but must check and then decide where to put them.

File: Lib/idlelib/config-extensions.def
Comment (terryjreedy) [inline Lib/idlelib/config-extensions.def:42]: Here you accomplished effect 1 above, `event_add(...)`, by adding the event-sequences pair to core bindings.  But this has the bug of adding then to the configure Keys tab.  Core bindings are configurable bindings.  I will add `event_add` elsewhere when I decide where.

File: Lib/idlelib/config-extensions.def
Comment (terryjreedy) [inline Lib/idlelib/config-extensions.def:54]: Ditto.

File: Lib/idlelib/config-extensions.def
Comment (terryjreedy) [inline Lib/idlelib/config-extensions.def:55]: This needs event_add.
The presence of <KeyRelease-0> was a bug that caused calltip to flash off and on.  I removed it.

File: Lib/idlelib/config-extensions.def
Comment (terryjreedy) [inline Lib/idlelib/config-extensions.def:65]: The is a dummy line as there is not key to associate with a virtual event.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
The fixed key bindings are not yet handled properly, resulting in multiple bugs.  Except for one detail, I believe I know what to do and will do so tomorrow, and then retest.

**Comment by terryjreedy** [comment]
This may be complete enough, but I may decide to change the new mac bindings first, as I really do not like to leave IDLE in a can't-release regressed state after a commit.

'Code context'  is visible on the option menu for Shell and Output Window but not operative.  It should be gone (as before) or grayed out, but the current situation could be tolerated in a release.

**Comment by terryjreedy** [comment]
I ran test suite a couple of times, but now there are two failures to look at.

FAIL: test_get_current_keyset (idlelib.idle_test.test_config.IdleConfTest)
Traceback (most recent call last):
  File "/home/travis/build/python/cpython/Lib/idlelib/idle_test/test_config.py", line 526, in test_get_current_keyset
    self.assertNotEqual(conf.GetCurrentKeySet(), conf.GetKeySet(conf.CurrentKeys()))

FAIL: test_click_help_source (idlelib.idle_test.test_query.HelpsourceGuiTest)
Traceback (most recent call last):
  File "C:\projects\cpython\lib\idlelib\idle_test\test_query.py", line 347, in test_click_help_source
    Equal(dialog.result, ('__test__', prefix + __file__))
AssertionError: Tuples differ: ('__test__', 'C:\\projects\\cpython\\lib\\idlelib\\idle[17 chars].py') != ('__test__', 'file://C:\\projects\\cpython\\lib\\idleli[24 chars].py')
First differing element 1:
'C:\\projects\\cpython\\lib\\idlelib\\idle_test\\test_query.py'
'file://C:\\projects\\cpython\\lib\\idlelib\\idle_test\\test_query.py'

**Comment by terryjreedy** [comment]
I ran tests before changing Mac keys.  I disabled test that depended on there being Alt keys in Mac or extension bindings.  The other failure was due to sys.platform not being restored in first test.

**Review by terryjreedy — APPROVED** [review]

**Comment by terryjreedy** [comment]
There are some regressions in the untested behavior of Config Dialog.  I suspect they occurred before this patch, but I need to check before merging.

**Comment by terryjreedy** [comment]
By checking out master, I verified that the problems I found yesterday pre-existed this patch.  The new one I found today required a fix to the new code.

**Pull request merged**
